src/main/com/yetanalytics/lrs/pedestal/interceptor/xapi/statements/attachment.cljc

```diff
@@ -133,145 +122,110 @@
+  Removes duplicate multiparts."
+  [multiparts]
+  (reduce (fn [m {:keys [headers] :as multipart}]
+            (let [sha (get headers "X-Experience-API-Hash")]
+              (if-let [extant (get m sha)]
+                m
+                (assoc m sha multipart))))
+          {}
+          multiparts))
 
-(defn validate-statements-multiparts
-  "Validate and return statements and their multipart attachments"
-  [statements multiparts]
-  (let [;; collect the attachments per statement and reduce over them
-        {valid-statements    :s-acc
-         valid-multiparts    :a-acc
-         leftover-multiparts :mps}
+(defn validate-multiparts
+  "Given a list of statements and a list of multiparts, return valid multiparts,
+  deduplicated."
+  [statements
+   multiparts]
+  (let [{:keys [sha2s
+                mpart-map]}
         (reduce
-         (fn [{:keys [mps] :as m}
-              [s & att-objs]]
-           ;; match the attachment objects to a multipart
-           ;; or assert that they are a fileURL
-           (let [;; Reduce over multiparts and find relevant attachments
-                 [next-mps
-                  valid-matched]
-                 (reduce
-                  (fn [[mps' acc :as state] ao]
-                    (if-let [[sha mp :as match]
-                             (make-sha-multipart-pair
-                              s mps' ao)]
-                      [(if (< 1 (count (get mps' sha)))
-                         (update mps' sha #(into [] (rest %)))
-                         (dissoc mps' sha))
-                       (conj acc match)]
-                      state))
-                  [mps []]
-                  att-objs)
-
-                 valid-shas    (map first valid-matched)]
-             (-> m
-                 (update :s-acc conj s)
-                 (update :a-acc into (map second valid-matched))
-                 (assoc :mps
-                        next-mps))))
-         {;; accumulators for the statements + attachments
-          :s-acc []
-          :a-acc []
-          ;; the multiparts to join up, by Hash
-          :mps   (multipart-map multiparts)}
-         (statements-attachments statements))]
-    (if (seq leftover-multiparts)
-      ;; If we have leftovers, it's bad
+         (fn [{:keys [sha2s mpart-map]
+               :as state}
+              {:keys [statement
+                      attachment-path]
+               {:strs [sha2 fileUrl]
+                :as att-obj} :attachment}]
+           (if-let [match-mp (get mpart-map sha2)]
+             (cond-> (update state :sha2s conj sha2)
+               ;; Validate + recreate sig mp
+               (and (= "attachments"
+                       (first attachment-path))
+                    (sig? att-obj))
+               (update-in
+                [:mpart-map sha2]
+                (partial validate-sig
+                         statement
+                         att-obj)))
```

Comment:
My gut says this block of code should be extracted into its own private helper function.

Reply:
https://github.com/yetanalytics/lrs/pull/81/commits/e4a6cd87204f9cdaaa2890e4e3dd6aeb1c6b8465